### ISSUE TITLE
Plural translation strings in JavaScript are not extracted

### ISSUE THREAD

**Issue by torfsen** [ckan/ckan#3151]
It seems that plural translation strings (using Jed's `isPlural` function) don't get extracted to `ckan.pot` when running `python setup.py extract_messages`.

Here's a small demo JS file:

``` javascript
this.ckan.module('message-extraction-test', function ($, _) {

  initialize: function() {
    console.log(_('Raw Singular Translation').fetch());
    console.log(_('Raw Plural Translation').isPlural(1, 'I can haz plural?').fetch());
    console.log(this.i18n('singular'));
    console.log(this.i18n('plural', {count: 2}));
  },

  options: {
    i18n: {
      singular: _('Nothing fancy'),
      plural: function (data) {
        return _('%(count)d item').isPlural(data.count, '%(count)d items');
      }
    }
  }
});
```

From that, `python setup.py extract_messages` extracts

```
#: ckan/message_extraction_test.js:4
msgid "Raw Singular Translation"
msgstr ""

#: ckan/message_extraction_test.js:5
msgid "Raw Plural Translation"
msgstr ""

#: ckan/message_extraction_test.js:12
msgid "Nothing fancy"
msgstr ""

#: ckan/message_extraction_test.js:14
#, python-format
msgid "%(count)d item"
msgstr ""
```

Note that `"I can haz plural?"` and `"%(count)d items"` are missing. The correct behavior would have been to recognize them as plural forms of `"Raw Plural Translation"` and `"%(count)d item"`, respectively, e.g.

```
#: ckan/message_extraction_test.js:5
msgid "Raw Plural Translation"
msgid_plural "I can haz plural?"
msgstr[0] ""
msgstr[1] ""
```

It seems that CKAN [configures Babel](https://github.com/ckan/ckan/blob/4bffd8c2611ae9eacec6b8593681fcc287d9a622/setup.cfg) to take `isPlural` into account:

```
[extract_messages]
keywords = translate isPlural
add_comments = TRANSLATORS:
output_file = ckan/i18n/ckan.pot
width = 80
```

However, that doesn't seem to be working.

**Comment by torfsen** [comment]
This might be a limitation of babel, see python-babel/babel#72.

**Comment by torfsen** [comment]
Even worse, the usual "gettext" approach of message extraction seems to assume that a singular form and the corresponding plural form are part of the same function call, see the description of `keywordspec` in the [xgettext documentation](https://www.gnu.org/software/gettext/manual/html_node/xgettext-Invocation.html). Babel [seems to use the same approach](https://github.com/python-babel/babel/blob/5fe694b57ef730c4154f77a78c5dbe819c6298a5/babel/messages/extract.py#L32), so Jed's `isPlural` might be difficult to support at all.

Instead of `isPlural` we could also use `ngettext`, which Jed also provides. However, that directly returns the string without interpolation (and not a Jed object with a `fetch` method):
```javascript
>> ckan.i18n.ngettext('%(count)d item', '%(count)d items', 2)
"%(count)d items"
```
Interpolation must hence be done manually using Jed's `sprintf`:
```javascript
>> var data = {count: 2};
>> var s = ckan.i18n.ngettext('%(count)d item', '%(count)d items', data.count);
>> ckan.i18n.sprintf(s, data)
"2 items"
```
As a fix I suggest the following: Instead of directly exposing a Jed instance, `ckan.i18n` should be a proxy that only offers the translation functions that we use in Python and Jinja2 (`_` and `ungettext`). That functions should also be available to JavaScript modules via `this._` and `this.ungettext`. This would have the benefit that translation calls look the same in Python, Jinja2 and JS.

This would mean that we get rid of the centralized translations in a module's `options.i18n` and the corresponding lookup-function `this.i18n`, but I don't see a problem with that.

I can prepare a PR for this, but first would like to get some feedback on whether the approach is OK.

**Comment by TkTech** [comment]
@torfsen I've thought it over, and can't think of a better solution than what you've described. Using a proxy method will also make the eventual transition to flask-babel a non-issue.

There are some changes we could make in babel to support isPlural in the future, but so far that discussion isn't going anywhere quickly.

**Comment by torfsen** [comment]
@TkTech thanks for your input! I'm currently working on other things but will start working on this once I find time.

**Comment by torfsen** [comment]
Somehow this issue didn't get closed when #3211 was merged, so I'm closing it now.
